### PR TITLE
MVP Cast API

### LINKED ISSUE #79: Cast API not implemented [$95]
This is probably one of the most critical components needed in order for most people to be able to switch away from binary gmscore, mainly because most media nowadays (i.e. youtube, netflix, etc.) depends on it in order to be at all useful. Everything else can be worked around, replaced, or proxied.

There is no real indicator of status of this, the closest is from 3rd party website, and indicates "Cast API: Some initial work done, nothing usable, applications may crash."
- what is the current status?
- what needs to be done?

### PR DESCRIPTION
**Pull request by armills**
## Description
This PR is the next step in the cast implementation, and is capable of playing videos with Google's sample Cast-Videos app. Basic cast functionality is now functional. Device discovery is implemented. Standard cast apps can now connect, start videos, and control playback.

Fixes #79

### Still broken
- There's still something preventing Netflix from starting videos. It must make use of some cast code that isn't yet implemented
- Volume control is not functional
- Joining an existing session is not implemented

### Dependencies
- [x] https://github.com/microg/android_external_GmsApi/pull/9
- [x] https://github.com/microg/android_external_SafeParcel/pull/3
- [x] https://github.com/microg/android_external_SafeParcel/pull/4

### CODE REVIEW

**Comment by armills** [comment]
OK, I've got a published branch working for the necessary upstream chromecast APIs. This PR builds and should be testable now.

**Comment by csagan5** [comment]
There are some new objects missing, probably additions in recent SDKs.

If I try to build Chromium by embedding various microg packages (not this branch), I get:
```
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:8: error: cannot find symbol
import com.google.android.gms.cast.MediaInfo;
                                  ^
  symbol:   class MediaInfo
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:9: error: cannot find symbol
import com.google.android.gms.cast.MediaStatus;
                                  ^
  symbol:   class MediaStatus
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:10: error: cannot find symbol
import com.google.android.gms.cast.RemoteMediaPlayer;
                                  ^
  symbol:   class RemoteMediaPlayer
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:11: error: package com.google.android.gms.cast.RemoteMediaPlayer does not exist
import com.google.android.gms.cast.RemoteMediaPlayer.MediaChannelResult;
                                                    ^
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:28: error: package RemoteMediaPlayer does not exist
public class RemoteMediaPlayerWrapper implements RemoteMediaPlayer.OnMetadataUpdatedListener,
                                                                  ^
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:29: error: package RemoteMediaPlayer does not exist
                                                 RemoteMediaPlayer.OnStatusUpdatedListener,
                                                                  ^
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:30: error: cannot find symbol
                                                 ResultCallback<MediaChannelResult>,
                                                                ^
  symbol: class MediaChannelResult
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:37: error: cannot find symbol
    private RemoteMediaPlayer mMediaPlayer;
            ^
  symbol:   class RemoteMediaPlayer
  location: class RemoteMediaPlayerWrapper
../../chrome/android/java/src/org/chromium/chrome/browser/media/remote/RemoteMediaPlayerWrapper.java:208: error: cannot find symbol
    public void onResult(MediaChannelResult result) {
                         ^
  symbol:   class MediaChannelResult
  location: class RemoteMediaPlayerWrapper
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/MediaStatusBridge.java:7: error: cannot find symbol
import com.google.android.gms.cast.MediaInfo;
                                  ^
  symbol:   class MediaInfo
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/MediaStatusBridge.java:8: error: cannot find symbol
import com.google.android.gms.cast.MediaMetadata;
                                  ^
  symbol:   class MediaMetadata
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/MediaStatusBridge.java:9: error: cannot find symbol
import com.google.android.gms.cast.MediaStatus;
                                  ^
  symbol:   class MediaStatus
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/MediaStatusBridge.java:20: error: cannot find symbol
    private MediaStatus mStatus;
            ^
  symbol:   class MediaStatus
  location: class MediaStatusBridge
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/MediaStatusBridge.java:22: error: cannot find symbol
    public MediaStatusBridge(MediaStatus status) {
                             ^
  symbol:   class MediaStatus
  location: class MediaStatusBridge
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/CastSessionUtil.java:8: error: cannot find symbol
import com.google.android.gms.cast.RemoteMediaPlayer;
                                  ^
  symbol:   class RemoteMediaPlayer
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/CastSessionUtil.java:26: error: cannot find symbol
            CastDevice castDevice, RemoteMediaPlayer mediaPlayer) {
                                   ^
  symbol:   class RemoteMediaPlayer
  location: class CastSessionUtil
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastOptionsProvider.java:10: error: cannot find symbol
import com.google.android.gms.cast.framework.OptionsProvider;
                                            ^
  symbol:   class OptionsProvider
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastOptionsProvider.java:11: error: cannot find symbol
import com.google.android.gms.cast.framework.SessionProvider;
                                            ^
  symbol:   class SessionProvider
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastOptionsProvider.java:16: error: cannot find symbol
public class CastOptionsProvider implements OptionsProvider {
                                            ^
  symbol: class OptionsProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastOptionsProvider.java:26: error: cannot find symbol
    public List<SessionProvider> getAdditionalSessionProviders(Context context) {
                ^
  symbol:   class SessionProvider
  location: class CastOptionsProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastSessionController.java:12: error: cannot find symbol
import com.google.android.gms.cast.framework.CastSession;
                                            ^
  symbol:   class CastSession
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastSessionController.java:32: error: cannot find symbol
    private CastSession mCastSession;
            ^
  symbol:   class CastSession
  location: class CastSessionController
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:13: error: cannot find symbol
import com.google.android.gms.cast.framework.CastSession;
                                            ^
  symbol:   class CastSession
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:14: error: cannot find symbol
import com.google.android.gms.cast.framework.SessionManagerListener;
                                            ^
  symbol:   class SessionManagerListener
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:38: error: cannot find symbol
        implements MediaRouteProvider, DiscoveryDelegate, SessionManagerListener<CastSession> {
                                                          ^
  symbol: class SessionManagerListener
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:38: error: cannot find symbol
        implements MediaRouteProvider, DiscoveryDelegate, SessionManagerListener<CastSession> {
                                                                                 ^
  symbol: class CastSession
../../chrome/android/java/src/com/google/android/gms/cast/Cast.java:27: error: package org.microg.gms.cast does not exist
import org.microg.gms.cast.CastApiBuilder;
                          ^
../../chrome/android/java/src/com/google/android/gms/cast/Cast.java:28: error: package org.microg.gms.cast does not exist
import org.microg.gms.cast.CastApiImpl;
                          ^
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastSessionController.java:67: error: cannot find symbol
    public CastSession getSession() {
           ^
  symbol:   class CastSession
  location: class CastSessionController
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastSessionController.java:111: error: cannot find symbol
    public void attachToCastSession(CastSession session) {
                                    ^
  symbol:   class CastSession
  location: class CastSessionController
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastSessionController.java:116: error: cannot find symbol
    public void detachFromCastSession(CastSession session) {
                                      ^
  symbol:   class CastSession
  location: class CastSessionController
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:192: error: cannot find symbol
    public final void onSessionStarting(CastSession session) {
                                        ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:197: error: cannot find symbol
    public void onSessionStartFailed(CastSession session, int error) {
                                     ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:202: error: cannot find symbol
    public void onSessionStarted(CastSession session, String sessionId) {
                                 ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:217: error: cannot find symbol
    public final void onSessionResumed(CastSession session, boolean wasSuspended) {
                                       ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:222: error: cannot find symbol
    public final void onSessionResuming(CastSession session, String sessionId) {}
                                        ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:225: error: cannot find symbol
    public final void onSessionResumeFailed(CastSession session, int error) {}
                                            ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:228: error: cannot find symbol
    public final void onSessionEnding(CastSession session) {
                                      ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:235: error: cannot find symbol
    public final void onSessionEnded(CastSession session, int error) {}
                                     ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafBaseMediaRouteProvider.java:238: error: cannot find symbol
    public final void onSessionSuspended(CastSession session, int reason) {
                                         ^
  symbol:   class CastSession
  location: class CafBaseMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastUtils.java:7: error: cannot find symbol
import com.google.android.gms.cast.framework.CastContext;
                                            ^
  symbol:   class CastContext
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafMediaRouteProvider.java:12: error: cannot find symbol
import com.google.android.gms.cast.framework.CastSession;
                                            ^
                                            ^
  symbol:   class CastSession
  location: package com.google.android.gms.cast.framework
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CastUtils.java:14: error: cannot find symbol
    public static CastContext getCastContext() {
                  ^
  symbol:   class CastContext
  location: class CastUtils
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/caf/CafMediaRouteProvider.java:149: error: cannot find symbol
    public void onSessionStarted(CastSession session, String sessionId) {
                                 ^
  symbol:   class CastSession
  location: class CafMediaRouteProvider
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/cast/CastSessionImpl.java:12: error: cannot find symbol
import com.google.android.gms.cast.MediaStatus;
                                  ^
  symbol:   class MediaStatus
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/cast/CastSessionImpl.java:13: error: cannot find symbol
import com.google.android.gms.cast.RemoteMediaPlayer;
                                  ^
  symbol:   class RemoteMediaPlayer
  location: package com.google.android.gms.cast
../../chrome/android/java/src/org/chromium/chrome/browser/media/router/cast/CastSessionImpl.java:72: error: cannot find symbol
    private RemoteMediaPlayer mMediaPlayer;
            ^
  symbol:   class RemoteMediaPlayer
  location: class CastSessionImpl
../../chrome/android/java/src/com/google/android/gms/cast/CastRemoteDisplay.java:26: error: package org.microg.gms.cast does not exist
import org.microg.gms.cast.CastRemoteDisplayApiBuilder;
                          ^
../../chrome/android/java/src/com/google/android/gms/cast/CastRemoteDisplay.java:27: error: package org.microg.gms.cast does not exist
import org.microg.gms.cast.CastRemoteDisplayApiImpl;
                          ^
../../chrome/android/java/src/org/microg/gms/common/GmsClient.java:32: error: cannot find symbol
import com.google.android.gms.common.internal.IGmsCallbacks;
                                             ^
  symbol:   class IGmsCallbacks
  location: package com.google.android.gms.common.internal
../../chrome/android/java/src/org/microg/gms/common/GmsClient.java:33: error: cannot find symbol
import com.google.android.gms.common.internal.IGmsServiceBroker;
                                             ^
  symbol:   class IGmsServiceBroker
  location: package com.google.android.gms.common.internal
../../chrome/android/java/src/org/microg/gms/common/GmsClient.java:60: error: cannot find symbol
    protected void onConnectedToBroker(IGmsServiceBroker broker, GmsCallbacks callbacks) throws RemoteException {
                                       ^
  symbol:   class IGmsServiceBroker
  location: class GmsClient<I>
  where I is a type-variable:
    I extends IInterface declared in class GmsClient
../../chrome/android/java/src/org/microg/gms/common/GmsClient.java:165: error: package IGmsCallbacks does not exist
    public class GmsCallbacks extends IGmsCallbacks.Stub {
                                                   ^
```

Part of these errors are due to the packages I have not yet added, but some of them (`package com.google.android.gms.cast.RemoteMediaPlayer does not exist`, `error: cannot find symbol
import com.google.android.gms.cast.MediaStatus;`) are genuine.

(This was just an experiment, proper embedding should go via a JAR, but it gives an idea of what is missing)

**Comment by armills** [comment]
@csagan5 Let's please keep this discussion limited to merging this into microG for now. This isn't meant to be 100% complete. Also could you please remove that giant error dump? It's not useful for this PR and is interrupting the thread here.

**Comment by csagan5** [comment]
@armills I suggest asking project admins to amend the comments next time, it avoids further interruptions of the comments flow.

**Comment by armills** [comment]
@mar-v-in Is there anything you need before the PR can be reviewed?

**Comment by CosmoJG** [comment]
Will this be merged soon? Also, @armills, how did you accomplish this?

**Comment by ale5000-git** [comment]
Probably in march.

Quote from MaR-V-iN:
> I am still alive, but just super busy.
> 
> Starting end of February, I set aside 20 hours a week to work solely on microG.
> 
> But please don't expect any update earlier than that, I have to ensure that there are no loose ends before I remove myself from some of the obligations I have right now, so probably will be even busier than before.
> 
> Thanks for your ongoing support everyone.

**Comment by ale5000-git** [comment]
@armills: Could you please resolve conflicts?

**Comment by armills** [comment]
:+1: All set

File: settings.gradle [hunk 1]
```diff
@@ -1,4 +1,4 @@
-include ':wearable-lib'
+include ':wearable-lib', ':safe-parcel'
```
Comment (mar-v-in) [inline settings.gradle]: This doesn't look right, there is no subfolder "safe-parcel" anymore.

**Comment by pedro-araujo** [comment]
I just compiled and tested this, I can connect to my TV wirelessly and play videos with audio through the TV. I just have one question: the colors seemed a bit washed off and the frame rate a bit choppy, the audio sounds normal though. Is this expected due to the early implementation of the Cast API?

**Comment by armills** [comment]
Thanks for testing this out! I know this branch hasn't gotten a lot of real world testing, so it's good to know it's working for you!

The video quality won't really have anything to do with this implementation. The cast API just passes a video URL to the chromecast, which then streams the video directly from the source.

**Comment by MrDodojo** [comment]
> Thanks for testing this out! I know this branch hasn't gotten a lot of real world testing, so it's good to know it's working for you!
> 
> The video quality won't really have anything to do with this implementation. The cast API just passes a video URL to the chromecast, which then streams the video directly from the source.

This means YouTube Casting will work again?

**Comment by MrDodojo** [comment]
> Thanks for testing this out! I know this branch hasn't gotten a lot of real world testing, so it's good to know it's working for you!
> 
> The video quality won't really have anything to do with this implementation. The cast API just passes a video URL to the chromecast, which then streams the video directly from the source.

Nice! you could probably ask the YouTube Vanced gang to test this

**Comment by ale5000-git** [comment]
@armills: Note => There must be `fix #79` in the commit to get the bounty.

**Comment by CFPrulez** [comment]
What is currently preventing the branch from being merged? Is there anything that has to be done before?

**Comment by MrDodojo** [comment]
> What is currently preventing the branch from being merged? Is there anything that has to be done before?
Devs probs busy

**Comment by ivanslf** [comment]
For an MVP, this seems awesome! I will test in my Chromecast 3 and give some more feedback.

**Comment by ale5000-git** [comment]
I don't know the code enough to tell it but is `<item>` expected to be [here](https://github.com/microg/android_packages_apps_GmsCore/commit/a80b263656e756c5563c454050ff685ccabaf973#diff-e6aaad41595d44e91fe25c681d01c1b0R113)?

**Comment by mar-v-in** [comment]
No, it shouldn't be there, will fix it right away.

**Comment by ale5000-git** [comment]
@armills: Did you cliaimed the bounty [here](https://www.bountysource.com/issues/30090937-cast-api-not-implemented)?

**Comment by armills** [comment]
Submitted!

**Pull request merged**